### DISCUSSION TITLE
Handling Consistent Entity IDs Across Multiple ExcaliburJS Instances for Multiplayer Game Development

### DISCUSSION THREAD

**Discussion by JumpLink** [excaliburjs/Excalibur#2695] [Q&A] [answered]
I am developing a multiplayer game that is intended to operate across multiple devices/browsers, as well as in a splitscreen mode within a single device/browser. For splitscreen, I am creating multiple engine instances (possibly related to #2622). 

Currently, I have to ignore the entity IDs, as they differ for each instance. However, even when running the engines in different browsers, it's possible for the IDs to diverge, for instance, when players navigate to different maps or trigger events in a different order.

If the IDs were consistent across every instance, it would simplify data synchronization.

At present, ExcaliburJS uses the following logic for IDs:

```ts
class Entity {
    private static _ID = 0;
    /**
     * The unique identifier for the entity
     */
    public id: number = Entity._ID++;
    // ...
}
```

To ensure entities have consistent IDs across different ExcaliburJS instances, it seems necessary to set the IDs manually. I am currently considering the best way to implement this.

Should I simply add a separate additional ID for each entity?.

Or would it be worthwhile to modify ExcaliburJS to, for example, allow adding an ID generator method to the engine that entities would use to generate their ID if one is provided? 

Or should I perhaps solve this in a completely different way?

**Comment by JumpLink** [comment, marked as answer]
Or maybe I just use the name property for this 🤔

**Reply by eonarheim** [reply]
@JumpLink Sorry this fell off my radar!

I'd definitely look at using the name property for this, it's free for any type of developer use!

Originally it was added so that the debug tools/logs could display something nice but could be used for anything. It doesn't make any uniqueness guarantees so that'll be up to you.

**Reply by JumpLink** [reply]
@eonarheim Thank you, 

I will probably need this in the future for tile entities from Tiled maps too, would you accept corresponding contributions to Excalibur or excalibur-tiled? For example to set the name according to a scheme like "[map-name]-[z]-[x]-[y]" or something else if it is cleaner, like a callback function so set the name, or so.

Or what do you think would be the best solution to my original question, taking into account that I would also be willing to make appropriate contributions?

(I'm not sure yet if I'll need this at all, maybe I can just use the coordinates of the tiles together directly as UUID and don't need to set a name for it, this is more an example in which direction this will go)

**Reply by eonarheim** [reply]
@JumpLink I'd definitely accept contributions in excalibur-tiled around the naming of entities. Are you thinking of having a name on every Tile? Do you need name id's explicitly or are the coordinates in a TileMap good enough to uniquely identify tiles?

Unfortunately the internals of excalibur rely on the current id method and format for things and it might be some effort to change (but more research is required). So I'd probably stick with the recommendation of using the `name` field with a UUID for now to coordinate across Excalibur instances and treat the builtin `.id` as a local id. That said I'm definitely open to exploring a pluggable ID strategy more.

I'm very curious what limits you'll run into using the name as your UUID, (we could add more name based facilities to assist too).

**Reply by JumpLink** [reply]
I think I can get by with map name + x + y + plane to uniquely identify the tiles, I'm not there yet though, but I'll let you know :)
